File: 2025/JuyoungMoon/week2/2025-06-15-longest-repeating-character-replacement.cpp
// Author: Juyoung Moon
// KDD LeetCode Study Week 2: Two Pointer & Sliding Window.
// https://github.com/juyomo/leetcode-study

// LeetCode #424.
// https://leetcode.com/problems/longest-repeating-character-replacement/

class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> frequencies;

        int maxLen = 0;
        int maxFreq = 0;
        int i = 0;

        for (int j = 0; j < s.size(); j++) {
            frequencies[s[j]]++;
            maxFreq = max(maxFreq, frequencies[s[j]]);

            if ((j - i + 1) - maxFreq > k) {
                frequencies[s[i]]--;
                i++;
            }

            maxLen = max(maxLen, j - i + 1);
        }

        return maxLen;
    }
};

File: 2025/Jaehyuk/2025-06-15-Longest-Repeating-Character-Replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        ans = 0
        l = 0
        freq = dict()
        maxFreq = 0
        for r in range(len(s)):
            freq[s[r]] = 1 + freq.get(s[r], 0)
            maxFreq = max(maxFreq, freq[s[r]])
            while (r - l + 1) - maxFreq > k:
                freq[s[l]] -= 1
                l += 1
            ans = max(ans, r - l + 1)
        return ans